src/tox/config/__init__.py

```diff
@@ -1389,48 +1452,16 @@ def make_envconfig(self, name, section, subs, config, replace=True):
             envpython=tc.get_envpython,
             **subs
         )
-        for env_attr in config._testenv_attr:
-            atype = env_attr.type
-            try:
-                if atype in (
-                    "bool",
-                    "float",
-                    "path",
-                    "string",
-                    "dict",
-                    "dict_setenv",
-                    "argv",
-                    "argvlist",
-                    "argv_install_command",
-                ):
-                    meth = getattr(reader, "get{}".format(atype))
-                    res = meth(env_attr.name, env_attr.default, replace=replace)
-                elif atype == "basepython":
-                    no_fallback = name in (config.provision_tox_env,)
-                    res = reader.getstring(
-                        env_attr.name,
-                        env_attr.default,
-                        replace=replace,
-                        no_fallback=no_fallback,
-                    )
-                elif atype == "space-separated-list":
-                    res = reader.getlist(env_attr.name, sep=" ")
-                elif atype == "line-list":
-                    res = reader.getlist(env_attr.name, sep="\n")
-                elif atype == "env-list":
-                    res = reader.getstring(env_attr.name, replace=False)
-                    res = tuple(_split_env(res))
-                else:
-                    raise ValueError("unknown type {!r}".format(atype))
-                if env_attr.postprocess:
-                    res = env_attr.postprocess(testenv_config=tc, value=res)
-            except tox.exception.MissingSubstitution as e:
-                tc._missing_subs[env_attr.name] = res = e
-            # On Python 2, exceptions are handled in __getattr__
-            if not six.PY2 or not isinstance(res, Exception):
-                setattr(tc, env_attr.name, res)
-            if atype in ("path", "string", "basepython"):
-                reader.addsubstitutions(**{env_attr.name: res})
+        reader._tc = tc
+        # Force postprocess as there is the expectation that attributes
+        # postprocess runs at startup.
+        # This is needed for setting up setenv, and also for tox plugins
+        # that rely on this assumption.
+        for env_attr in self.config._testenv_attr.values():
+            if env_attr.postprocess:
```

Comment:
Note that removing this condition makes it 100% non-lazy again, after which this has the primary benefit of `_replace_substitution` using the new `_get_value` which has access to the `VenvAttribute` metadata for each variable.  The reason this can be helpful is that e.g. `install_command` should declare that `{opts}` and `{packages}` are "fake" variables, and this is needed inside the replacement engine so that they are re-escaped variables for each pass through the replacement engine.